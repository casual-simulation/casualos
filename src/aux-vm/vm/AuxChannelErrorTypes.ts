export type AuxChannelErrorType = AuxGeneralErrorType;

export interface AuxGeneralErrorType {
    type: 'general';
    message: string;
}
