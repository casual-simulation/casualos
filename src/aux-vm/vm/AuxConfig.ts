import {
    AuxPartitionConfig,
    hasValue,
    RemoteCausalRepoProtocol,
    SharedPartitionsVersion,
    StoredAuxVersion1,
} from '@casual-simulation/aux-common';
import { AuxDevice } from '@casual-simulation/aux-runtime';

/**
 * Defines the possible configuration options for a simulation.
 */
export interface AuxConfig {
    /**
     * The ID of the config bot.
     */
    configBotId: string;

    /**
     * The configuration parameters for the simulation.
     */
    config: AuxConfigParameters;

    /**
     * Defines the partitioning structure for bots.
     */
    partitions: AuxPartitionConfig;
}

export interface AuxConfigParameters {
    version: string | null;
    versionHash: string | null;
    builder?: string;
    bootstrapState?: StoredAuxVersion1;
    device?: AuxDevice;

    /**
     * Whether to only allow compiling and runing scripts
     * that have valid signatures.
     */
    forceSignedScripts?: boolean;

    /**
     * Whether to run in debug mode.
     * This will likely cause more verbose console output.
     */
    debug?: boolean;

    /**
     * The connection protocol that causal repo partitions should use.
     */
    causalRepoConnectionProtocol?: RemoteCausalRepoProtocol;

    /**
     * The URL that causal repo partitions should connect to.
     */
    causalRepoConnectionUrl?: string;

    /**
     * The version of the shared partitions that should be used.
     */
    sharedPartitionsVersion?: SharedPartitionsVersion;

    /**
     * The origin that the VM iframe should be loaded from.
     */
    vmOrigin?: string;

    /**
     * The origin that the Auth iframe should be loaded from.
     */
    authOrigin?: string;

    /**
     * The origin that the records API is hosted at.
     */
    recordsOrigin?: string;

    /**
     * The list of portal bots that should be automatically created
     * before the sync event is triggered.
     */
    builtinPortals?: string[];

    /**
     * The configuration needed to perform timesync for the simulation.
     */
    timesync?: AuxTimeSyncConfiguration;

    /**
     * Gets the player mode of this CasualOS version.
     *
     * - "player" indicates that the instance has been configured for experiencing AUXes.
     * - "builder" indicates that the instance has been configured for building AUXes.
     */
    playerMode?: 'player' | 'builder';
}

export interface AuxTimeSyncConfiguration {
    /**
     * The host that timesync requests should be made to.
     */
    host?: string;

    /**
     * The protocol that should be used for time sync requests.
     */
    connectionProtocol?: TimeSyncProtocol;
}

export type TimeSyncProtocol = RemoteCausalRepoProtocol;

export function buildVersionNumber(config: AuxConfigParameters) {
    if (!config) {
        return null;
    }
    return {
        hash: config.versionHash,
        ...parseVersionNumber(config.version),
        playerMode: config.playerMode ?? 'builder',
    };
}

/**
 * Parses the given version number.
 * @param version The version number.
 */
export function parseVersionNumber(version: string) {
    if (!version) {
        return {
            version: null,
            major: null,
            minor: null,
            patch: null,
            alpha: null,
        };
    }
    const versionRegex = /^v(\d+)\.(\d+)\.(\d+)((\:|-)\w+\.?\d*)*$/i;
    const [str, major, minor, patch, prerelease] = versionRegex.exec(version);

    let alpha: boolean | number = false;
    if (hasValue(prerelease)) {
        alpha = true;
        const [first, number] = prerelease.split('.');
        if (hasValue(number)) {
            alpha = parseInt(number);
        }
    }

    return {
        version: str,
        major: parseInt(major),
        minor: parseInt(minor),
        patch: parseInt(patch),
        alpha,
    };
}
