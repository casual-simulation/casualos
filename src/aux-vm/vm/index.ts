export * from './AuxConfig';
export * from './AuxChannel';
export * from './AuxVM';
export * from './AuxHelper';
export * from './BaseAuxChannel';
export * from './AuxChannelErrorTypes';
