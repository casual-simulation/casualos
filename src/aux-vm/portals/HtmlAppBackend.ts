import {
    action,
    asyncResult,
    hasValue,
    ON_APP_SETUP_ACTION_NAME,
    registerHtmlApp,
    SerializableMutationRecord,
    updateHtmlApp,
    unregisterHtmlApp,
    htmlAppMethod,
} from '@casual-simulation/aux-common';
import { AuxHelper } from '../vm';
import { AppBackend } from './AppBackend';
import { v4 as uuid } from 'uuid';
import undom, {
    BUILTIN_HTML_ELEMENT_FUNCTIONS,
    registerMethodHandler,
    RootNode,
    supressMutations,
} from '@casual-simulation/undom';
import { render } from 'preact';
import { BehaviorSubject, Subscription } from 'rxjs';
import { first, map } from 'rxjs/operators';
import { RuntimeActions } from '@casual-simulation/aux-runtime';

export const TARGET_INPUT_PROPERTIES = ['value', 'checked'];

/**
 * Properties that should automatically be copied for specific tag types.
 * For performance, these properties are only copied if an event is sent for the element with the element as the target.
 */
export const ELEMENT_SPECIFIC_PROPERTIES: { [nodeName: string]: string[] } = {
    IMG: ['width', 'height', 'naturalWidth', 'naturalHeight', 'currentSrc'],
    VIDEO: ['videoWidth', 'videoHeight', 'duration', 'currentSrc'],
    SECTION: ['scrollTop', 'offsetHeight'],
};

const TEXT_REFERENCE_PROPERTIES = ['data'];
const NODE_REFERENCE_PROPERTIES = [
    'namespace',
    'nodeName',
    'style',
    'attributes',
    'className',
];

export interface HtmlPortalSetupResult {
    builtinEvents: string[];
}

/**
 * Defines a class that is used to communicate HTML changes for a custom html portal.
 */
export class HtmlAppBackend implements AppBackend {
    appId: string;
    botId: string;

    private _helper: AuxHelper;
    private _sub: Subscription;
    private _initTaskId: string;
    private _registerTaskId: string | number;
    private _instanceId: string;
    private _document: Document;
    private _mutationObserver: MutationObserver;
    private _nodes: Map<string, RootNode> = new Map<string, RootNode>();
    private _initialContent: any;
    private _setupObservable: BehaviorSubject<boolean>;
    private _methodCallTasks: Map<number | string, (result: any) => void> =
        new Map();

    /**
     * the list of properties that should be disallowed.
     * Taken from https://github.com/developit/preact-worker-demo/blob/master/src/renderer/worker.js
     */
    private _propDenylist: Set<string> = new Set([
        'children',
        'parentNode',
        '__handlers',
        '_component',
        '_componentConstructor',
        'l',
    ]);

    private _propCopyList: Set<string> = new Set(['style']);

    /**
     * The list of properties that should be converted to references.
     * Taken from https://github.com/developit/preact-worker-demo/blob/master/src/renderer/worker.js
     */
    private _propReferenceList: Set<string> = new Set([
        'target',
        'addedNodes',
        'removedNodes',
        'nextSibling',
        'previousSibling',
    ]);

    private _idCounter = 0;

    get onSetup() {
        return this._setupObservable.pipe(
            first((setup) => !!setup),
            map(() => {})
        );
    }

    get document() {
        return this._document;
    }

    constructor(
        appId: string,
        botId: string,
        helper: AuxHelper,
        registerTaskId?: string | number,
        instanceId?: string
    ) {
        this.appId = appId;
        this.botId = botId;
        this._registerTaskId = registerTaskId;
        this._sub = new Subscription();
        this._sub.add(() => {
            this._renderContent('');
            this._helper.transaction(
                unregisterHtmlApp(this.appId, this._instanceId)
            );
        });

        this._helper = helper;
        this._setupObservable = new BehaviorSubject(false);

        this._initTaskId = uuid();
        this._instanceId = instanceId ?? uuid();
        this._helper.transaction(
            registerHtmlApp(this.appId, this._instanceId, this._initTaskId)
        );
    }

    handleEvents(events: RuntimeActions[]): void {
        for (let event of events) {
            if (event.type === 'async_result') {
                if (event.taskId === this._initTaskId) {
                    this._setupApp(event.result as HtmlPortalSetupResult);
                } else if (this._methodCallTasks.has(event.taskId)) {
                    this._methodCallTasks.get(event.taskId)(event.result);
                }
            } else if (event.type === 'html_app_event') {
                if (event.appId === this.appId) {
                    let target = this._getNode(event.event.target);
                    if (target && target.dispatchEvent) {
                        let finalEvent = {
                            ...event.event,
                            target: target,
                            bubbles: true,
                        };

                        try {
                            supressMutations(true);
                            for (let prop of TARGET_INPUT_PROPERTIES) {
                                let eventPropName = `_target${prop}`;
                                if (eventPropName in finalEvent) {
                                    (<any>target)[prop] =
                                        finalEvent[eventPropName];
                                }
                            }
                            const propList =
                                ELEMENT_SPECIFIC_PROPERTIES[target.nodeName];
                            if (propList) {
                                for (let prop of propList) {
                                    let eventPropName = `_target${prop}`;
                                    if (eventPropName in finalEvent) {
                                        (<any>target)[prop] =
                                            finalEvent[eventPropName];
                                    }
                                }
                            }
                        } finally {
                            supressMutations(false);
                        }

                        try {
                            target.dispatchEvent(finalEvent);
                        } catch (err) {
                            console.error(err);
                        }
                    }
                }
            } else if (event.type === 'set_app_output') {
                if (event.appId === this.appId) {
                    if (typeof event.output === 'object') {
                        if (!this._document) {
                            this._initialContent = event.output;
                            continue;
                        }

                        this._renderContent(event.output);
                    }
                }
            }
        }
    }

    dispose(): void {
        this._sub.unsubscribe();
    }

    private _renderContent(content: any) {
        let prevDocument = globalThis.document;
        try {
            globalThis.document = this._document;
            render(content, this._document.body);
        } catch (err) {
            console.error(err);
        } finally {
            globalThis.document = prevDocument;
        }
    }

    private _getNode(node: any): RootNode {
        let id: string;
        if (node && typeof node === 'object') {
            id = node.__id;
        } else if (typeof node === 'string') {
            id = node;
        }
        if (!id) {
            return null;
        }

        if (node.nodeName === 'BODY') {
            return this._document.body as any;
        }
        return this._nodes.get(id);
    }

    private _setupApp(result: HtmlPortalSetupResult) {
        try {
            let doc = (this._document = undom({
                builtinEvents: result?.builtinEvents,
            }));

            this._registerMethodHandlers(doc);

            this._mutationObserver = new doc.defaultView.MutationObserver(
                this._processMutations.bind(this)
            );
            this._mutationObserver.observe(doc, {
                subtree: true,
            });

            this._helper.transaction(
                action(ON_APP_SETUP_ACTION_NAME, [this.botId], undefined, {
                    document: this._document,
                })
            );

            if (this._initialContent) {
                this._renderContent(this._initialContent);
            }

            if (hasValue(this._registerTaskId)) {
                this._helper.transaction(
                    asyncResult(this._registerTaskId, null)
                );
            }
        } finally {
            this._setupObservable.next(true);
        }
    }

    private _registerMethodHandlers(doc: Document) {
        for (let method of BUILTIN_HTML_ELEMENT_FUNCTIONS) {
            this._registerVoidMethodHandler(doc, 'HTMLElement', method);
        }

        const inputElemenetFunctions = [
            'select',
            'setCustomValidity',
            'setRangeText',
            'setSelectionRange',
            'showPicker',
            'stepDown',
            'stepUp',
        ];
        for (let method of inputElemenetFunctions) {
            this._registerVoidMethodHandler(doc, 'HTMLInputElement', method);
        }

        const formElementFunctions = ['reset', 'submit'];
        for (let method of formElementFunctions) {
            this._registerPromiseMethodHandler(doc, 'HTMLFormElement', method);
        }

        const mediaElementVoidFunctions = ['fastSeek', 'load', 'pause'];
        const mediaElementPromiseFunctions = ['play'];

        for (let method of mediaElementPromiseFunctions) {
            this._registerPromiseMethodHandler(doc, 'HTMLMediaElement', method);
        }

        for (let method of mediaElementVoidFunctions) {
            this._registerVoidMethodHandler(doc, 'HTMLMediaElement', method);
        }

        const videoElementFunctions = [
            // 'getVideoPlaybackQuality',
            'requestPictureInPicture',
        ];
        for (let method of videoElementFunctions) {
            this._registerPromiseMethodHandler(doc, 'HTMLVideoElement', method);
        }
    }

    private _registerVoidMethodHandler(
        doc: Document,
        className: string,
        methodName: string
    ) {
        this._sub.add(
            registerMethodHandler(
                doc,
                className,
                methodName,
                (el, method, args) => {
                    this._emitMethodCall(el, methodName, args);
                    return undefined;
                }
            )
        );
    }

    private _registerPromiseMethodHandler(
        doc: Document,
        className: string,
        methodName: string
    ) {
        this._sub.add(
            registerMethodHandler(
                doc,
                className,
                methodName,
                (el, method, args) => {
                    return this._emitMethodCall(el, methodName, args);
                }
            )
        );
    }

    private _emitMethodCall(
        element: any,
        methodName: string,
        args: any[]
    ): Promise<any> {
        return new Promise<any>((resolve, reject) => {
            const taskId = uuid();

            this._methodCallTasks.set(taskId, (result) => {
                this._methodCallTasks.delete(taskId);
                resolve(result);
            });

            try {
                this._helper.transaction(
                    htmlAppMethod(
                        this.appId,
                        this._getNodeId(element),
                        methodName,
                        args,
                        taskId
                    )
                );
            } catch (err) {
                reject(err);
            }
        });
    }

    private _processMutations(mutations: MutationRecord[]) {
        for (let mutation of mutations) {
            for (let prop of this._propReferenceList) {
                (<any>mutation)[prop] = this._makeReference(
                    (<any>mutation)[prop]
                );
            }
        }

        this._helper.transaction(updateHtmlApp(this.appId, mutations as any[]));
    }

    private _getNodeId(obj: RootNode) {
        let id = (<any>obj).__id;
        if (!id) {
            id = (<any>obj).__id = (this._idCounter++).toString();
        }
        this._nodes.set(id, obj);
        return id;
    }

    // Mostly copied from https://github.com/developit/preact-worker-demo/blob/bac36d7c34b241e4c041bcbdefaef77bcc5f367e/src/renderer/worker.js#L81
    private _makeReference(obj: RootNode | RootNode[]): any {
        if (!obj || typeof obj !== 'object') {
            return obj;
        }

        if (Array.isArray(obj)) {
            return obj.map(this._makeReference, this);
        }

        const anyObj = obj as any;
        let id = anyObj.__id;
        if (!id) {
            id = anyObj.__id = (this._idCounter++).toString();
        }
        this._nodes.set(id, obj);

        let result = {
            __id: id,
            nodeType: obj.nodeType,
        } as any;

        if (obj.nodeType === 3) {
            for (let prop of TEXT_REFERENCE_PROPERTIES) {
                result[prop] = anyObj[prop];
            }
        } else if (obj.nodeType === 1) {
            for (let prop of NODE_REFERENCE_PROPERTIES) {
                if (!prop.startsWith('_') || prop === '__id') {
                    const value = anyObj[prop];
                    if (hasValue(value)) {
                        if (this._propCopyList.has(prop)) {
                            result[prop] = { ...value };
                        } else {
                            result[prop] = value;
                        }
                    }
                }
            }

            if (anyObj.childNodes) {
                result.childNodes = this._makeReference(anyObj.childNodes);
            }
        }

        return result;
    }
}
