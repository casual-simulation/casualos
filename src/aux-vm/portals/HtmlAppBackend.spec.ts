import {
    asyncResult,
    AuxPartitions,
    BotAction,
    botAdded,
    createBot,
    createMemoryPartition,
    htmlAppEvent,
    iteratePartitions,
    MemoryPartition,
    ON_APP_SETUP_ACTION_NAME,
    registerHtmlApp,
    setAppOutput,
    stateUpdatedEvent,
    toast,
    unregisterHtmlApp,
    updateHtmlApp,
    UpdateHtmlAppAction,
} from '@casual-simulation/aux-common';
import { AuxRuntime, RuntimeActions } from '@casual-simulation/aux-runtime';
import { waitAsync } from '@casual-simulation/aux-common/test/TestHelpers';
import { Subscription } from 'rxjs';
import { HtmlAppBackend } from './HtmlAppBackend';
import { v4 as uuid } from 'uuid';
import { AuxHelper } from '../vm';
import { skip, tap } from 'rxjs/operators';
import { h } from 'preact';
import htm from 'htm';
import { htmlAppMethod } from '@casual-simulation/aux-common/bots/BotEvents';

const uuidMock: jest.Mock = <any>uuid;
jest.mock('uuid');

describe('HtmlAppBackend', () => {
    let runtime: AuxRuntime;
    let actions: RuntimeActions[];
    let memory: MemoryPartition;
    let userId: string = 'user';
    let sub: Subscription;

    beforeEach(async () => {
        actions = [];
        sub = new Subscription();
        runtime = new AuxRuntime(
            {
                hash: 'hash',
                major: 1,
                minor: 0,
                patch: 0,
                version: 'v1.0.0',
                alpha: true,
                playerMode: 'builder',
            },
            {
                supportsAR: false,
                supportsVR: false,
                isCollaborative: true,
                allowCollaborationUpgrade: true,
                ab1BootstrapUrl: 'ab1Bootstrap',
            }
        );
        memory = createMemoryPartition({
            type: 'memory',
            initialState: {},
        });
        memory.space = 'shared';

        await memory.applyEvents([botAdded(createBot('user'))]);
    });

    function createHelper(partitions: AuxPartitions) {
        runtime = new AuxRuntime(
            {
                hash: 'hash',
                major: 1,
                minor: 0,
                patch: 0,
                version: 'v1.0.0',
                alpha: true,
                playerMode: 'builder',
            },
            {
                supportsAR: false,
                supportsVR: false,
                isCollaborative: true,
                allowCollaborationUpgrade: true,
                ab1BootstrapUrl: 'ab1Bootstrap',
            }
        );
        const helper = new AuxHelper(userId, partitions, runtime);

        for (let [, partition] of iteratePartitions(partitions)) {
            sub.add(
                partition.onStateUpdated
                    .pipe(
                        tap((e) => {
                            runtime.stateUpdated(e);
                        })
                    )
                    .subscribe({ error: (e: any) => console.error(e) })
            );
        }

        runtime.userId = userId;
        sub.add(helper.localEvents.subscribe((a) => actions.push(...a)));

        return helper;
    }

    afterEach(() => {
        sub.unsubscribe();
    });

    it('should send a register_html_app event when created', async () => {
        const helper = createHelper({
            shared: memory,
        });
        uuidMock.mockReturnValueOnce('uuid');

        let portal = new HtmlAppBackend(
            'testPortal',
            'myBot',
            helper,
            undefined,
            'appId'
        );

        await waitAsync();

        expect(actions).toEqual([
            registerHtmlApp('testPortal', 'appId', 'uuid'),
        ]);
    });

    describe('onAppSetup', () => {
        it('should send a onRender action when the register_html_app result is returned', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@tags.rendered = true`,
                        rendered: false,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend('testPortal', 'myBot', helper);

            await waitAsync();

            expect(helper.botsState['myBot'].tags.rendered).toBe(false);

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            expect(runtime.currentState['myBot'].values.rendered).toBe(true);
        });

        it('should include the HTML document in the onAppSetup action', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@that.document.body.appendChild(that.document.createElement('h1'))`,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                undefined,
                'appId'
            );

            await waitAsync();

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            expect(actions).toEqual([
                registerHtmlApp('testPortal', 'appId', 'uuid'),
                updateHtmlApp('testPortal', [
                    {
                        type: 'childList',
                        target: expect.objectContaining({
                            __id: '0',
                        }),
                        addedNodes: [
                            expect.objectContaining({
                                __id: '1',
                                nodeName: 'H1',
                            }),
                        ],
                        removedNodes: [],
                        attributeName: undefined,
                        attributeNamespace: undefined,
                        nextSibling: undefined,
                        previousSibling: undefined,
                        oldValue: undefined,
                    },
                ]),
            ]);
        });

        it('should call event listeners for the html_app_event events', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@let h1 = that.document.createElement('h1'); h1.addEventListener('click', () => tags.clicked = true); that.document.body.appendChild(h1);`,
                        clicked: false,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend('testPortal', 'myBot', helper);

            await waitAsync();

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            portal.handleEvents([
                htmlAppEvent('testPortal', {
                    type: 'click',
                    target: '1',
                }),
            ]);

            await waitAsync();

            expect(runtime.currentState['myBot'].values.clicked).toBe(true);
        });

        it('should emit an async result for the given register task ID', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@os.toast("Hit")`,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                'taskId'
            );

            await waitAsync();

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            expect(actions.slice(1)).toEqual([
                toast('Hit'),
                asyncResult('taskId', null),
            ]);
        });

        it('should trigger the onSetup observable', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(botAdded(createBot('myBot', {})));

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend('testPortal', 'myBot', helper);
            let setup = false;
            portal.onSetup.subscribe(() => (setup = true));

            await waitAsync();

            expect(setup).toBe(false);

            portal.handleEvents([asyncResult('uuid', null)]);

            await waitAsync();

            expect(setup).toBe(true);
        });

        it('should issue a html_app_method_call when .focus() is called', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(
                botAdded(
                    createBot('myBot', {
                        [ON_APP_SETUP_ACTION_NAME]: `@os.toast(that.document.body.focus())`,
                    })
                )
            );

            uuidMock.mockReturnValueOnce('uuid1').mockReturnValueOnce('uuid2');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                undefined,
                'appId'
            );

            await waitAsync();

            portal.handleEvents([asyncResult('uuid1', null)]);

            await waitAsync();

            expect(actions).toEqual([
                registerHtmlApp('testPortal', 'appId', 'uuid1'),
                htmlAppMethod('testPortal', '0', 'focus', [], 'uuid2'),
                toast(undefined),
            ]);
        });
    });

    describe('set_app_output', () => {
        it('should render the output as a preact component', async () => {
            const helper = createHelper({
                shared: memory,
            });

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                undefined,
                'appId'
            );

            await waitAsync();

            const html = htm.bind(h);

            portal.handleEvents([asyncResult('uuid', null)]);
            portal.handleEvents([
                setAppOutput('testPortal', html`<h1>Hello</h1>`),
            ]);

            await waitAsync();

            expect(actions.length).toBe(2);
            expect(actions[0]).toEqual(
                registerHtmlApp('testPortal', 'appId', 'uuid')
            );

            const updateAction = actions[1] as UpdateHtmlAppAction;

            expect(updateAction).toMatchSnapshot();

            expect(updateAction.type).toBe('update_html_app');
            expect(updateAction.appId).toBe('testPortal');
            expect(updateAction.updates.length).toBe(1);
            expect(updateAction.updates[0].type).toBe('childList');
            expect(updateAction.updates[0].addedNodes.length).toBe(1);
            const h1Node = updateAction.updates[0].addedNodes[0] as any;
            expect(h1Node.nodeName).toBe('H1');
            expect(h1Node.childNodes.length).toBe(1);
            expect(h1Node.childNodes[0].nodeType).toBe(3);
        });

        it('should render the most recent output when the app is setup', async () => {
            const helper = createHelper({
                shared: memory,
            });

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                undefined,
                'appId'
            );

            await waitAsync();

            const html = htm.bind(h);

            portal.handleEvents([
                setAppOutput('testPortal', html`<h1>Hello</h1>`),
                asyncResult('uuid', null),
            ]);

            await waitAsync();

            expect(actions.length).toBe(2);
            expect(actions[0]).toEqual(
                registerHtmlApp('testPortal', 'appId', 'uuid')
            );

            const updateAction = actions[1] as UpdateHtmlAppAction;

            expect(updateAction).toMatchSnapshot();

            expect(updateAction.type).toBe('update_html_app');
            expect(updateAction.appId).toBe('testPortal');
            expect(updateAction.updates.length).toBe(1);
            expect(updateAction.updates[0].type).toBe('childList');
            expect(updateAction.updates[0].addedNodes.length).toBe(1);
            const h1Node = updateAction.updates[0].addedNodes[0] as any;
            expect(h1Node.nodeName).toBe('H1');
            expect(h1Node.childNodes.length).toBe(1);
            expect(h1Node.childNodes[0].nodeType).toBe(3);
        });

        it('should not replace input elements when setting the value attribute', async () => {
            const helper = createHelper({
                shared: memory,
            });

            uuidMock.mockReturnValueOnce('uuid1').mockReturnValueOnce('uuid2');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                undefined,
                'appId'
            );

            await waitAsync();

            const html = htm.bind(h);

            const listener1 = jest.fn();
            const listener2 = jest.fn();

            portal.handleEvents([
                setAppOutput(
                    'testPortal',
                    h('div', {}, h('input', { value: '', onInput: listener1 }))
                ),
                asyncResult('uuid1', null),
            ]);

            await waitAsync();

            expect(actions.length).toBe(2);
            expect(actions[0]).toEqual(
                registerHtmlApp('testPortal', 'appId', 'uuid1')
            );

            const updateAction = actions[1] as UpdateHtmlAppAction;

            expect(updateAction).toMatchSnapshot();

            expect(updateAction.type).toBe('update_html_app');
            expect(updateAction.appId).toBe('testPortal');
            expect(updateAction.updates.length).toBe(2);
            expect(updateAction.updates[0].type).toBe('event_listener');
            expect(updateAction.updates[0].listenerName).toBe('input');
            expect(updateAction.updates[1].type).toBe('childList');
            expect(updateAction.updates[1].addedNodes.length).toBe(1);
            const inputNode1 = updateAction.updates[1].addedNodes[0] as any;
            expect(inputNode1.nodeName).toBe('DIV');

            portal.handleEvents([
                setAppOutput(
                    'testPortal',
                    h(
                        'div',
                        {},
                        h('input', { value: 'my value', onInput: listener1 })
                    )
                ),
                asyncResult('uuid2', null),
            ]);

            await waitAsync();

            const updateAction2 = actions[2] as UpdateHtmlAppAction;

            expect(updateAction2).toMatchSnapshot();

            expect(updateAction2.type).toBe('update_html_app');
            expect(updateAction2.appId).toBe('testPortal');
            expect(updateAction2.updates.length).toBe(1);
            expect(updateAction2.updates[0].type).toBe('attributes');
            expect(updateAction2.updates[0].attributeName).toBe('value');
            expect((<any>updateAction2.updates[0].target).attributes).toEqual([
                { name: 'value', value: 'my value', ns: null },
            ]);
        });
    });

    describe('dispose()', () => {
        it('should send a unregister_html_app event', async () => {
            const helper = createHelper({
                shared: memory,
            });
            await helper.transaction(botAdded(createBot('myBot', {})));

            uuidMock.mockReturnValueOnce('uuid');

            let portal = new HtmlAppBackend(
                'testPortal',
                'myBot',
                helper,
                undefined,
                'appId'
            );

            await waitAsync();

            portal.dispose();

            await waitAsync();

            expect(actions.slice(1)).toEqual([
                unregisterHtmlApp('testPortal', 'appId'),
            ]);
        });
    });
});
