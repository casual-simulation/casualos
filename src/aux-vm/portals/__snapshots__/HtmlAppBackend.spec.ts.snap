// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlAppBackend set_app_output should not replace input elements when setting the value attribute 1`] = `
Object {
  "appId": "testPortal",
  "type": "update_html_app",
  "updates": Array [
    Object {
      "addedNodes": undefined,
      "listenerDelta": 1,
      "listenerName": "input",
      "nextSibling": undefined,
      "previousSibling": undefined,
      "removedNodes": undefined,
      "target": null,
      "type": "event_listener",
    },
    Object {
      "addedNodes": Array [
        Object {
          "__id": "1",
          "attributes": Array [],
          "childNodes": Array [
            Object {
              "__id": "2",
              "attributes": Array [
                Object {
                  "name": "value",
                  "ns": null,
                  "value": "",
                },
              ],
              "childNodes": Array [],
              "nodeName": "INPUT",
              "nodeType": 1,
              "style": Object {},
            },
          ],
          "nodeName": "DIV",
          "nodeType": 1,
          "style": Object {},
        },
      ],
      "nextSibling": undefined,
      "previousSibling": undefined,
      "removedNodes": Array [],
      "target": Object {
        "__id": "0",
        "attributes": Array [],
        "childNodes": Array [
          Object {
            "__id": "1",
            "attributes": Array [],
            "childNodes": Array [
              Object {
                "__id": "2",
                "attributes": Array [
                  Object {
                    "name": "value",
                    "ns": null,
                    "value": "",
                  },
                ],
                "childNodes": Array [],
                "nodeName": "INPUT",
                "nodeType": 1,
                "style": Object {},
              },
            ],
            "nodeName": "DIV",
            "nodeType": 1,
            "style": Object {},
          },
        ],
        "nodeName": "BODY",
        "nodeType": 1,
        "style": Object {},
      },
      "type": "childList",
    },
  ],
  Symbol(UNMAPPABLE): true,
}
`;

exports[`HtmlAppBackend set_app_output should not replace input elements when setting the value attribute 2`] = `
Object {
  "appId": "testPortal",
  "type": "update_html_app",
  "updates": Array [
    Object {
      "addedNodes": undefined,
      "attributeName": "value",
      "attributeNamespace": null,
      "nextSibling": undefined,
      "oldValue": "",
      "previousSibling": undefined,
      "removedNodes": undefined,
      "target": Object {
        "__id": "2",
        "attributes": Array [
          Object {
            "name": "value",
            "ns": null,
            "value": "my value",
          },
        ],
        "childNodes": Array [],
        "nodeName": "INPUT",
        "nodeType": 1,
        "style": Object {},
      },
      "type": "attributes",
    },
  ],
  Symbol(UNMAPPABLE): true,
}
`;

exports[`HtmlAppBackend set_app_output should render the most recent output when the app is setup 1`] = `
Object {
  "appId": "testPortal",
  "type": "update_html_app",
  "updates": Array [
    Object {
      "addedNodes": Array [
        Object {
          "__id": "1",
          "attributes": Array [],
          "childNodes": Array [
            Object {
              "__id": "2",
              "data": "Hello",
              "nodeType": 3,
            },
          ],
          "nodeName": "H1",
          "nodeType": 1,
          "style": Object {},
        },
      ],
      "nextSibling": undefined,
      "previousSibling": undefined,
      "removedNodes": Array [],
      "target": Object {
        "__id": "0",
        "attributes": Array [],
        "childNodes": Array [
          Object {
            "__id": "1",
            "attributes": Array [],
            "childNodes": Array [
              Object {
                "__id": "2",
                "data": "Hello",
                "nodeType": 3,
              },
            ],
            "nodeName": "H1",
            "nodeType": 1,
            "style": Object {},
          },
        ],
        "nodeName": "BODY",
        "nodeType": 1,
        "style": Object {},
      },
      "type": "childList",
    },
  ],
  Symbol(UNMAPPABLE): true,
}
`;

exports[`HtmlAppBackend set_app_output should render the output as a preact component 1`] = `
Object {
  "appId": "testPortal",
  "type": "update_html_app",
  "updates": Array [
    Object {
      "addedNodes": Array [
        Object {
          "__id": "1",
          "attributes": Array [],
          "childNodes": Array [
            Object {
              "__id": "2",
              "data": "Hello",
              "nodeType": 3,
            },
          ],
          "nodeName": "H1",
          "nodeType": 1,
          "style": Object {},
        },
      ],
      "nextSibling": undefined,
      "previousSibling": undefined,
      "removedNodes": Array [],
      "target": Object {
        "__id": "0",
        "attributes": Array [],
        "childNodes": Array [
          Object {
            "__id": "1",
            "attributes": Array [],
            "childNodes": Array [
              Object {
                "__id": "2",
                "data": "Hello",
                "nodeType": 3,
              },
            ],
            "nodeName": "H1",
            "nodeType": 1,
            "style": Object {},
          },
        ],
        "nodeName": "BODY",
        "nodeType": 1,
        "style": Object {},
      },
      "type": "childList",
    },
  ],
  Symbol(UNMAPPABLE): true,
}
`;
